### PR TITLE
[pigeon] Documentation comments

### PR DESCRIPTION
**Pull request by tarrinneal**
Adds the ability to write documentation comments on pigeon inputs and have them generated in the output files.

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes https://github.com/flutter/flutter/issues/110207

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by tarrinneal** [comment]
Needs tests still

**Comment by tarrinneal** [comment]
Let me know if you think I should add more tests for this.

File: packages/pigeon/test/cpp_generator_test.dart
Comment (stuartmorgan) [inline packages/pigeon/test/cpp_generator_test.dart:1122]: It would be good to have these (for all of the generators) check for the actual comment format, to ensure in the unit tests that they have been correctly comment-ified.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
This is great! Just a few minor comments. Mostly inline below, but also capturing from offline discussion: it would be good to have a file in `pigeons/` that has one of every type of comment, and for that to be generated and compiled for every target language. That way if, say, someone later made a change that broke compilation (like forgetting a `*/` in some generated output) it would be caught by tests.

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [inline packages/pigeon/lib/ast.dart:60]: I'd add to the docstring why this is a List.  Is this all the lines of the docstring?

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [inline packages/pigeon/lib/ast.dart:94]:
```suggestion
  /// Any attached docummentation comments on the API.
```

Comment (gaaclarke) [inline packages/pigeon/lib/ast.dart:94]: Same comment about why this is a list.

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [inline packages/pigeon/lib/ast.dart:183]:
```suggestion
  /// Comments documenting the NamedType.
```

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [inline packages/pigeon/lib/ast.dart:207]:
```suggestion
  /// List of documentation comments.
```

Comment (gaaclarke) [inline packages/pigeon/lib/ast.dart:207]: Same feedback about List.

File: packages/pigeon/lib/ast.dart
Comment (gaaclarke) [inline packages/pigeon/lib/ast.dart:227]:
```suggestion
  /// List of documentation comments.
```

File: packages/pigeon/lib/cpp_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/cpp_generator.dart:751]: This structure is duplicated across the code.  I'd factor out a helper function, `_writeDocstring`?

File: packages/pigeon/lib/dart_generator.dart
Comment (gaaclarke) [inline packages/pigeon/lib/dart_generator.dart:575]: This too is duplicated.  I'd factor this out to a helper function too.

**Review by gaaclarke — COMMENTED** [review]
Looks like @stuartmorgan already noticed the same thing about duplicating logic.  Looks great overall, nice work.

File: packages/pigeon/lib/ast.dart
Comment (tarrinneal) [reply packages/pigeon/lib/ast.dart:61]: I debated writing it like that instead, and decided against it. I'll go ahead and make those changes though.

File: packages/pigeon/lib/cpp_generator.dart
Comment (tarrinneal) [reply packages/pigeon/lib/cpp_generator.dart:11]: Sounds good, I just went with the style the code was already using. I'll go ahead and clean up the old comments as well.

File: packages/pigeon/lib/generator_tools.dart
Comment (gaaclarke) [inline packages/pigeon/lib/generator_tools.dart:437]: nit: this would be easier to follow with named parameters since sometimes one of the arguments is `null` for example.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (gaaclarke) [inline packages/pigeon/lib/pigeon_lib.dart:883]: stray print statement?

**Review by gaaclarke — COMMENTED** [review]
lgtm modulo stray print

File: packages/pigeon/lib/pigeon_lib.dart
Comment (tarrinneal) [reply packages/pigeon/lib/pigeon_lib.dart:883]: good catch, not sure how the formatter missed that

File: packages/pigeon/lib/generator_tools.dart
Comment (tarrinneal) [reply packages/pigeon/lib/generator_tools.dart:437]: Good point

File: packages/pigeon/CHANGELOG.md
Comment (stuartmorgan) [inline packages/pigeon/CHANGELOG.md:1]: Nit: 4.1.0 since this is new functionality (per [semver](https://semver.org/)).

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:11]: For C++ this should just be `//`; see https://google.github.io/styleguide/cppguide.html#Comment_Style

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:275]: If you also rename the constant to `commentPrefix` you could use that everywhere you're currently using `//` directly.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:462]: Here and in several places below you are using `///` directly instead of the constant.

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:15]: Nit: How about "docCommentPrefix"? "Open" implies that there should be a "close", which isn't the case for this comment style.

File: packages/pigeon/lib/generator_tools.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:436]: I think it's worth elaborating here about `open` and `close`, providing the examples of what they would be for `///` and `/** */`-style comments. Also, consider calling them `openCommentToken` and `closeCommentToken` so they are more self-documenting.

File: packages/pigeon/lib/generator_tools.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:438]: Similarly, it would be helpful to have a paragraph in the documentation comment explaining `additionalComments` vs `comments`, with an example of what the former will do.

Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:438]: `List<String> additionalComments = const <String>[]` (per the earlier comment about not using nullable collections)

File: packages/pigeon/lib/generator_tools.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:439]: Note that this is a bug caused by allowing nullable collections, because you aren't returning if `additionalComments` is empty but non-null.

File: packages/pigeon/lib/generator_tools.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:443]: Similar bug here.

File: packages/pigeon/lib/generator_tools.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:444]: Could you collapse this whole condition to `comments.length + additionalComments.length > 1`?

File: packages/pigeon/lib/generator_tools.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/generator_tools.dart:469]: In fact, I think you could probably simplify all of this logic quite a bit if you start with:

```
List<String> allComments = <String>[
  ...comments,
  if (comments.isNotEmpty && additionalComments.isNotEmpty) '',
  ...additionalComments,
];
```

and then operate only on `allComments`.

File: packages/pigeon/lib/cpp_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/cpp_generator.dart:11]: In all the files, these constants should either be `_`-prefixed, or annotated with `@visibleForTesting`, depending on whether you need them in tests.

File: packages/pigeon/lib/generator_tools.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/generator_tools.dart:436]: I think you'll also need a `continuationToken` (or similar name), since the standard way of writing the `/** */` multi-line comment is with a `*` on each new line (see https://google.github.io/styleguide/javaguide.html#s7.1-javadoc-formatting for an example).

File: packages/pigeon/pigeons/message.dart
Comment (stuartmorgan) [inline packages/pigeon/pigeons/message.dart:27]: At least some documentation should be multi-line for manual testing of that case; a class would be a good candidate. So:

```
/// This comment is to test class documentation comments.
///
/// This one has multiple lines.
```

File: packages/pigeon/test/java_generator_test.dart
Comment (stuartmorgan) [inline packages/pigeon/test/java_generator_test.dart:1144]: Can you use a raw string (`r''`) to simplify the escaping here? (I thought there was a lint about that.)

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
The structure looks great now, just some local comments.

File: packages/pigeon/test/java_generator_test.dart
Comment (tarrinneal) [reply packages/pigeon/test/java_generator_test.dart:1144]: That would require me to add string concatenation to replace the $comment. Not sure which is the greater evil in this case.

Comment (tarrinneal) [reply packages/pigeon/test/java_generator_test.dart:1144]: I've updated it, let me know if you think it should be reverted, or if you know of a better way.

Comment (stuartmorgan) [reply packages/pigeon/test/java_generator_test.dart]: Ah, I missed that or I wouldn't have flagged it. I still think it's better this way even with the string concatenation, but it could go either way so I'll leave it up to you which one you want to go with.

**Review by stuartmorgan — APPROVED** [review]
LGTM with a few small suggestions. If there's anything you want me to take another look at after you make those last changes feel free to re-request review, but it's not necessary.

File: packages/pigeon/lib/generator_tools.dart
Comment (tarrinneal) [reply packages/pigeon/lib/generator_tools.dart:458]: lol, oops.

File: packages/pigeon/lib/pigeon_lib.dart
Comment (stuartmorgan) [reply packages/pigeon/lib/pigeon_lib.dart:899]: Good point 🙂  I was stuck on the idea of accepting either form and then outputting with good spacing, but yes, preserving whatever spacing the developer did (or didn't) add is definitely the better easy option.

File: packages/pigeon/lib/dart_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/dart_generator.dart:15]: Optional: you could just inline this since we don't expect to ever use a documentation comment marker except for with `addDocumentationComments` (unlike the C++ version where it's the same as a normal comment).

File: packages/pigeon/lib/java_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:17]: Same here; you could inline these.

File: packages/pigeon/lib/java_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/java_generator.dart:14]: Shouldn't this not have a trailing space now?

File: packages/pigeon/lib/swift_generator.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/swift_generator.dart:10]: Same optional nit.

File: packages/pigeon/lib/ast.dart
Comment (stuartmorgan) [inline packages/pigeon/lib/ast.dart:95]: I think it's worth fleshing this out with discussion of the non-obvious whitespace thing. E.g.:

```
/// List of documentation comments, separated by line.
///
/// Lines should not include the comment marker itself, but should include any
/// leading whitespace, so that any indentation in the original comment is preserved.
/// For example: [" List of documentation comments, separated by line.", ...]
```
```

**Review by stuartmorgan — APPROVED** [review]
LGTM with nits (most of them optional)

File: packages/pigeon/lib/dart_generator.dart
Comment (tarrinneal) [reply packages/pigeon/lib/dart_generator.dart:15]: I was going for consistency between generator files. I'll defer to you on preference though.

Comment (stuartmorgan) [reply packages/pigeon/lib/dart_generator.dart:15]: > I was going for consistency between generator files.

Makes sense.

**Pull request merged**
